source/common/config/ads_api_impl.cc

```diff
@@ -0,0 +1,161 @@
+  }
+  async_client_.reset(new Grpc::AsyncClientImpl<envoy::api::v2::DiscoveryRequest,
+                                                envoy::api::v2::DiscoveryResponse>(
+      cluster_manager, ads_config.cluster_name()[0]));
+}
+
+AdsApiImpl::~AdsApiImpl() {
+  for (auto watches : watches_) {
+    for (auto watch : watches.second) {
+      watch->inserted_ = false;
+    }
+  }
+}
+
+void AdsApiImpl::start() {
+  if (async_client_) {
+    establishNewStream();
+  }
+}
+
+void AdsApiImpl::setRetryTimer() {
+  retry_timer_->enableTimer(std::chrono::milliseconds(RETRY_DELAY_MS));
+}
+
+void AdsApiImpl::establishNewStream() {
+  // TODO(htuch): stats
+  ENVOY_LOG(debug, "Establishing new gRPC bidi stream for {}", service_method_.DebugString());
+  stream_ = async_client_->start(service_method_, *this);
+  if (stream_ == nullptr) {
+    ENVOY_LOG(warn, "Unable to establish new stream");
+    handleFailure();
+    return;
+  }
+
+  for (const auto type_url : subscriptions_) {
+    sendDiscoveryRequest(requests_[type_url]);
+  }
+}
+
+void AdsApiImpl::sendDiscoveryRequest(const envoy::api::v2::DiscoveryRequest& request) {
+  if (stream_ == nullptr) {
+    return;
+  }
+  stream_->sendMessage(request, false);
+}
+
+void AdsApiImpl::handleFailure() {
+  for (auto watches : watches_) {
+    for (auto watch : watches.second) {
+      watch->callbacks_.onConfigUpdateFailed(nullptr);
+    }
+  }
+  setRetryTimer();
+}
+
+AdsWatchPtr AdsApiImpl::subscribe(const std::string& type_url,
+                                  const std::vector<std::string>& resources,
+                                  AdsCallbacks& callbacks) {
+  auto watch =
+      std::unique_ptr<AdsWatch>(new AdsWatchImpl(resources, callbacks, watches_[type_url]));
+  ENVOY_LOG(debug, "ADS subscribe for " + type_url);
+
+  // Lazily kick off the requests based on first subscription. This has the
+  // convenient side-effect that we order messages on the channel based on
+  // Envoy's internal dependency ordering.
+  if (requests_.count(type_url) == 0) {
+    requests_[type_url].set_type_url(type_url);
+    requests_[type_url].mutable_node()->CopyFrom(node_);
+    subscriptions_.push_front(type_url);
+    sendDiscoveryRequest(requests_[type_url]);
+  }
+
+  return watch;
+}
+
+void AdsApiImpl::onCreateInitialMetadata(Http::HeaderMap& metadata) {
+  UNREFERENCED_PARAMETER(metadata);
+}
+
+void AdsApiImpl::onReceiveInitialMetadata(Http::HeaderMapPtr&& metadata) {
+  UNREFERENCED_PARAMETER(metadata);
+}
+
+void AdsApiImpl::onReceiveMessage(std::unique_ptr<envoy::api::v2::DiscoveryResponse>&& message) {
+  const std::string& type_url = message->type_url();
+  ENVOY_LOG(debug, "Received ADS message for {}", type_url);
+  try {
+    // To avoid O(n^2) explosion (e.g. when we have 1000s of EDS watches), we
+    // build a map here from resource name to resource and then walk watches_.
+    // TODO(htuch): Reconsider implementation data structure for watches to make lookups of resource
```

Comment:
yeah, why not just store internally as map?

Reply:
We expect that most watches that name resources (as opposed to open CDS/LDS watches, which just wildcard via an empty resource list) will have a single resource, but the current APIs internally allows for more. We need to assemble the subset of resources from those received and pass them to `onConfigUpdate()` once per watch. This means we need to do some gather operation. This could be either done by building buckets for each watch and populating them as we walk the watch list (as done today), or building buckets and populating them as we walk the resource list.

I'm not a fan of either of these. If we change the API to allow zero (wildcard) or one resource per watch, we can easily use a map. Given that's how it's actually used by the xDS implementations in Envoy today, I think I'm going to modify the ADS internal API to switch to this. Let me know if that sounds good.

Reply:
+1 sounds good simpler is better

Reply:
After playing around with this, we still need to do the map construction. This is because we need to deliver empty updates to watches (e.g. EDS) when a resource is not present, so we basically need to walk all the watches and have a O(1) map back to the delivered resource matching the watched name. So, I'll leave it as is.